java/ql/src/experimental/Security/CWE/CWE-200/AndroidFileIntentSink.qll

```diff
@@ -0,0 +1,64 @@
+/** Provides Android sink models related to file creation. */
+
+import java
+import semmle.code.java.dataflow.DataFlow
+import semmle.code.java.dataflow.ExternalFlow
+import semmle.code.java.frameworks.android.Android
+import semmle.code.java.frameworks.android.Intent
+
+/** A sink representing methods creating a file in Android. */
+class AndroidFileSink extends DataFlow::Node {
+  AndroidFileSink() { sinkNode(this, "create-file") }
+}
+
+/**
+ * The Android class `android.os.AsyncTask` for running tasks off the UI thread to achieve
+ * better user experience.
+ */
+class AsyncTask extends RefType {
+  AsyncTask() { this.hasQualifiedName("android.os", "AsyncTask") }
+}
+
+/** The `execute` method of Android `AsyncTask`. */
+class AsyncTaskExecuteMethod extends Method {
+  AsyncTaskExecuteMethod() {
+    this.getDeclaringType().getSourceDeclaration().getASourceSupertype*() instanceof AsyncTask and
+    this.getName() = "execute"
+  }
+
+  int getParamIndex() { result = 0 }
```

Comment:
Does anything use these getParamIndex methods?

Reply:
It's not being used now. I modelled it to be used in the following query:
https://github.com/github/codeql/blob/0fa7d234ec0d6589b4917e76ac74ff9daedb61d8/java/ql/src/experimental/Security/CWE/CWE-200/SensitiveAndroidFileLeak.ql#L42-L46

Using these methods, the query becomes the following, which isn't concise so I simply used the index 0 or 1 directly:
```
        aema.getMethod() instanceof AsyncTaskExecuteMethod and
        prev.asExpr() = aema.getArgument(aema.getMethod().(AsyncTaskExecuteMethod).getParamIndex())
        or
        aema.getMethod() instanceof AsyncTaskExecuteOnExecutorMethod and
        prev.asExpr() = aema.getArgument(aema.getMethod().(AsyncTaskExecuteOnExecutorMethod).getParamIndex())
```

I will change the implementation to the following:
```
/** The method that executes `AsyncTask` of Android. */
abstract class ExecuteAsyncTaskMethod extends Method {
  /** Returns index of the parameter that is tainted. */
  abstract int getParamIndex();
}

/** The `execute` method of Android `AsyncTask`. */
class AsyncTaskExecuteMethod extends ExecuteAsyncTaskMethod {
  ...

  override int getParamIndex() { result = 0 }
}

/** The `executeOnExecutor` method of Android `AsyncTask`. */
class AsyncTaskExecuteOnExecutorMethod extends ExecuteAsyncTaskMethod {
  ...

  override int getParamIndex() { result = 1 }
}


      aema.getQualifier().getType() = arm.getDeclaringType() and
      aema.getMethod() instanceof ExecuteAsyncTaskMethod and
      prev.asExpr() = aema.getArgument(aema.getMethod().(ExecuteAsyncTaskMethod).getParamIndex()) and
      succ.asExpr() = arm.getParameter(0).getAnAccess()

```

Reply:
Either way don't include dead code in this PR; add it in the one that uses it.

Reply:
Added a new abstract class <code>ExecuteAsyncTaskMethod</code> and invoked the method.